crates/chia-client/src/peer.rs

```diff
@@ -1,363 +1,341 @@
+    /// Sends a message to the peer and expects a specific response message.
+    pub async fn request_infallible<T, B>(&self, body: B) -> Result<T>
     where
-        Response: Streamable + ChiaProtocolMessage,
         T: Streamable + ChiaProtocolMessage,
+        B: Streamable + ChiaProtocolMessage,
     {
         let message = self.request_raw(body).await?;
-        let data = message.data.as_ref();
-
-        if message.msg_type == Response::msg_type() {
-            Response::from_bytes(data).or(Err(Error::InvalidResponse(message)))
-        } else {
-            Err(Error::InvalidResponse(message))
+        if message.msg_type != T::msg_type() {
+            return Err(Error::InvalidResponse(
+                vec![T::msg_type()],
+                message.msg_type,
+            ));
         }
+        Ok(T::from_bytes(&message.data)?)
     }
 
-    pub async fn request_raw<T, R>(&self, body: T) -> Result<Message, Error<R>>
+    /// Sends a message to the peer and expects any arbitrary protocol message without parsing it.
+    pub async fn request_raw<T>(&self, body: T) -> Result<Message>
     where
         T: Streamable + ChiaProtocolMessage,
     {
-        // Get the current nonce and increment.
-        let message_id = self.nonce.fetch_add(1, Ordering::SeqCst);
+        let (sender, receiver) = oneshot::channel();
 
-        // Create the message.
         let message = Message {
             msg_type: T::msg_type(),
-            id: Some(message_id),
-            data: stream(&body)?.into(),
-        };
-
-        // Create a saved oneshot channel to receive the response.
-        let (sender, receiver) = oneshot::channel::<Message>();
-        self.requests.lock().await.insert(message_id, sender);
-
-        // Send the message.
-        let bytes = match stream(&message) {
-            Ok(bytes) => bytes.into(),
-            Err(error) => {
-                self.requests.lock().await.remove(&message_id);
-                return Err(error.into());
-            }
-        };
-        let send_result = self.sink.lock().await.send(bytes).await;
-
-        if let Err(error) = send_result {
-            self.requests.lock().await.remove(&message_id);
-            return Err(error.into());
+            id: Some(self.0.requests.insert(sender).await),
+            data: body.to_bytes()?.into(),
         }
+        .to_bytes()?
+        .into();
 
-        // Wait for the response.
-        let response = receiver.await;
-
-        // Remove the one shot channel.
-        self.requests.lock().await.remove(&message_id);
-
-        // Handle the response, if present.
-        response.or(Err(Error::MissingResponse))
+        self.0.sink.lock().await.send(message).await?;
+        Ok(receiver.await?)
     }
+}
 
-    pub fn receiver(&self) -> &broadcast::Receiver<PeerEvent> {
-        &self.event_receiver
+impl Drop for PeerInner {
+    fn drop(&mut self) {
+        self.inbound_handle.abort();
```

Comment:
this looks a bit suspicious. If we haven't joined this coroutine by the time this is dropped, isn't it too late?

Reply:
If you drop an `AbortHandle` to a task, the default behavior is to leave it running in the background. The more sensible behavior here is to just kill the task.

That said, I think the underlying `Stream` might close if the `Sink` is destroyed, so this might not do much.